plugins/arm-lifter/arm_main.ml

```diff
@@ -0,0 +1,239 @@
+      !%(cmprsr src1 src2 shift_reg shift_imm cond)
+    | `CMPrsi, [|src1; src2; shift_imm; cond; _|] ->
+      !%(cmprsi src1 src2 shift_imm cond)
+    | `CMNri, [|src1; src2; cond; _|] ->
+      !%(cmnri src1 src2 cond)
+    | `CMNzrr, [|src1; src2; cond; _|] ->
+      !%(cmnzrr src1 src2 cond)
+    | `CMNzrsr, [|src1; src2; shift_reg; shift_imm; cond; _|] ->
+      !%(cmnzrsr src1 src2 shift_reg shift_imm cond)
+    | `CMNzrsi, [|src1; src2; shift_imm; cond; _|] ->
+      !%(cmnzrsi src1 src2 shift_imm cond)
+    (* Special Data Instructions *)
+    | `MOVi16, [|`Reg dest; src; cond; _wflag|] ->
+      !%(movi16 (`Reg dest) src cond _wflag)
+    | `MOVTi16, [|`Reg dest; _; src; cond; _wflag|] ->
+      !%(movti16 (`Reg dest) src cond _wflag)
+    | _, _ -> raise (Defs.Lift_Error "unrecognized move instruction variant")
+
+  let lift_bits insn ops =
+    match insn, ops with
+    | _, _ -> pass
+
+  let lift_mult insn ops =
+    match insn,ops with
+    | _, _ -> pass
+
+  let lift_mem insn ops =
+    match insn, ops with
+    | _, _ -> pass
+
+  (** Branching instructions *)
+
+  let lift_branch insn ops =
+    match insn, ops with
+    | _, _ -> skip
+
+  let lift_special insn ops =
+    match insn, ops with
+    | _, _ -> pass
+
+  let lift_with (addr : Bitvec.t) (insn : Arm_defs.insn)
+      (ops : Arm_defs.op array) = match insn with
+    | #move_insn -> lift_move insn ops addr
+    | #mem_insn -> lift_mem insn ops |> move
+    | #bits_insn -> lift_bits insn ops |> move
+    | #mult_insn -> lift_mult insn ops |> move
+    | #special_insn -> lift_special insn ops |> move
+    (* this is malformed for now *)
+    | #branch_insn -> ctrl (lift_branch insn ops) pass (Bitvec.zero)
+
+end
+
+open Bap.Std
+
+let run_lifter _label addr insn _mem 
+    (lifter : Bitvec.t -> Defs.insn -> Defs.op array -> unit Theory.eff) =
+  match Insns.of_basic insn with
+  | None -> raise (Defs.Lift_Error "unknown instruction")
+  | Some arm_insn -> 
+    match Insns.arm_ops (Disasm_expert.Basic.Insn.ops insn) with
+    | Error err -> raise (Defs.Lift_Error (Error.to_string_hum err))
+    | Ok ops -> lifter addr arm_insn ops
+
+let provide_sematics insn effect = 
+  KB.provide Theory.Program.Semantics.slot insn effect >>| fun () -> insn
+
+let lift label =
+  Theory.instance () >>= Theory.require >>= fun (module Core) ->
+  KB.collect Theory.Label.addr label >>= fun addr -> (* this is the address of the current instruction *)
+  KB.collect Disasm_expert.Basic.Insn.slot label >>= fun insn -> (* the LLVM provided decoding *)
+  KB.collect Memory.slot label >>= fun mem -> (* the memory chunk, probably not needed *)
+  let module Lifter = ARM(Core) in
+  match addr, insn, mem with
+  | Some addr, Some insn, Some mem -> 
+    run_lifter label addr insn mem Lifter.lift_with >>= fun sema ->
+    provide_sematics label sema
+  | _ -> raise (Defs.Lift_Error "insufficient knowledge") (* should I have not raised an error here? *)
```

Comment:
no-no, no raises, the lifter shall never raise an exception in case if it is unable to provide knowledge. It is totally normal not to know something. Basically, the idea is that each lifter is providing as much knowledge as it can, but no more. Also, the simple rule in BAP for exceptions - exceptions are for indicating programmer errors. 

TL;DR; when there is not enough information to provide the semantics, just return the bottom value of the semantics, e.g., `Insn.empty` which is the short alias for `Knowledge.Value.empty Theory.Semantics.cls`.

Reply:
Is there a way to embed the error message into some kinds of empty knowledge like `BIL.special` do ?

> TL;DR; when there is not enough information to provide the semantics, just return the bottom value of the semantics, e.g., `Insn.empty` which is the short alias for `Knowledge.Value.empty Theory.Semantics.cls`.

By the way, speaking of `BIL.special`, how do I (or is it necessary) to express the same sematics as `BIL.cpuexn` ?